### PR TITLE
JENKINS-72213 : Add a waiting spinner during parameter building

### PR DESCRIPTION
**Pull request by c3p0-maif**
This PR show a waiting spinner at the beginning of update parameter method and hide it at the end.  
I am not totally satisfy about the spinner position (other item's positions are impacted by visibility changes) but my js-knowledge is not good enough to make it better.  

I use this pipeline to test the wanted comportement.  
```groovy
properties([
  parameters([
    [
      name: 'States',
      $class: 'ChoiceParameter',
      choiceType: 'PT_SINGLE_SELECT',
      description: 'Choose a state',
      script: [
        $class: 'GroovyScript',
        script: [sandbox:false, script: '''return ['Sao Paulo','Rio de Janeiro','Parana','Acre']''']
      ]
    ],
    [
      name: 'Cities',
      $class: 'CascadeChoiceParameter',
      choiceType: 'PT_SINGLE_SELECT',
      referencedParameters: 'States',
      description: 'Choose a city',
      script: [
        $class: 'GroovyScript',
        script: [sandbox:false, script: '''if (States == "Sao Paulo") {
  sleep(1000)
  return ["Barretos", "Sao Paulo", "Itu"]
} else if (States == "Rio de Janeiro") {
  sleep(2000)
  return ["Rio de Janeiro", "Mangaratiba"]
} else if (States == "Parana") {
  sleep(3000)
  return ["Curitiba", "Ponta Grossa"]
} else if (States == "Acre") {
  sleep(4000)
  return ["Rio Branco", "Acrelandia"]
} else {
  sleep(10000)
  return ["Unknown state"]
}''']
        ]
   ]
 ])
])


pipeline {
    agent any

    stages {
        stage('Hello') {
            steps {
                script {
                    echo "Hello from ${params.Cities}, ${params.States}"
                }
            }
        }
    }
}
```

Feel free to add your improvements.

### Testing done

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Review by kinow — COMMENTED** [review]
Thank you so much! It worked in previous versions of Jenkins, but the spinner/loading indicator stopped working, I think, about the time there was a tables-to-div big re-write.

I tried to find what was broken but had no luck. May I ask where/how did you find about what had to be fixed here? :nerd_face: 

Code-wise it looks good, will check out the branch locally later and review. If it works I will merge it and try to release it over next day(s). Thank you a lot @c3p0-maif !!!

**Comment by c3p0-maif** [comment]
@kinow, Thank you for your feedback.
I wasn't aware that this feature have already been present in this plugin, my PR came from a selfish need 😄 
However, this feature deserves to be reviewed by a js expert because i'm not totally satisfy about the final renderer.

**Comment by kinow** [comment]
Not an expert but I can try to review and give some feedback. Thanks again!!!

**Review by kinow — COMMENTED** [review]
One test is failing on `master`, I suspect due to some issue with the jquery3 adjunt plugin and Jenkins BOM (:sighs).

Another test is failing for me after I check out this branch. There might be some test that needs to be adjusted due to the new HTML spinner element.

@c3p0-maif, maybe the spinner should block the HTML element? So that users cannot change it while it is being updated?

What it looks like for me:


https://github.com/jenkinsci/active-choices-plugin/assets/304786/f2f9d96d-7227-459d-a18f-63a544efe40f

File: src/main/resources/org/biouno/unochoice/DynamicReferenceParameter/index.jelly
Comment (c3p0-maif) [reply src/main/resources/org/biouno/unochoice/DynamicReferenceParameter/index.jelly:71]: Hi @kinow,
It's because in dynamic reference parameters the paramName value is changed by the subelement (input, orderedList, formattedHtml,... ex : [here](https://github.com/jenkinsci/active-choices-plugin/blob/master/src/main/resources/org/biouno/unochoice/DynamicReferenceParameter/formattedHtml.jelly#L7) )
As this is this element which trigger the update method, I must change the id

**Comment by c3p0-maif** [comment]
> @c3p0-maif, maybe the spinner should block the HTML element? So that users cannot change it while it is being updated?
> 
> What it looks like for me:
> 
>  2023-11-19.14-16-42.mp4

Hi @kinow ,
Yes "simple" components should be blocked during the update but on custom HTML parameter it will be difficult

**Comment by c3p0-maif** [comment]
Hi @kinow , any news from this PR ?

**Comment by kinow** [comment]
Sorry, I'm having a look at the TAP plug-in, but will come back to this one after that.

**Comment by kinow** [comment]
One test on `master` passes on GitHub Actions but fails for me. Three tests on this branch fails for me, and apparently on GitHub Actions too (it just says "failed: 3", I didn't check if the same tests). I think it's just a matter of making the tests pass, doing some quick tests with parameters, and then this can be quickly merged and released :+1:

**Comment by c3p0-maif** [comment]
> One test on `master` passes on GitHub Actions but fails for me. Three tests on this branch fails for me, and apparently on GitHub Actions too (it just says "failed: 3", I didn't check if the same tests). I think it's just a matter of making the tests pass, doing some quick tests with parameters, and then this can be quickly merged and released 👍

No idea from my side how to manage this :(

**Comment by kinow** [comment]
Not a problem, I will figure out what's wrong, fix the build locally, test this PR and merge/release (as soon as I find time).

**Comment by kinow** [comment]
Note to self, here's the error:

```bash
[ERROR] Errors: 
[ERROR]   UiAcceptanceTest.test:163 » ElementClickIntercepted element click intercepted: Element <input name="value" alt="B" json="B" type="checkbox" title="B" value="B" class=" "> is not clickable at point (378, 712). Other element would receive the click: <button class="jenkins-button jenkins-button--primary jenkins-!-build-color">...</button>
  (Session info: chrome=121.0.6167.139)
Build info: version: '4.17.0', revision: 'e52b1be057*'
System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.15.0-92-generic', java.version: '17.0.9'
Driver info: org.openqa.selenium.chrome.ChromeDriver
Command: [828b2d6e917a0aca2f6e986d388255db, clickElement {id=A448766197E080F3F774D7A455CEF83C_element_23}]
Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 121.0.6167.139, chrome: {chromedriverVersion: 121.0.6167.85 (3f98d690ad7e..., userDataDir: /tmp/.org.chromium.Chromium...}, fedcm:accounts: true, goog:chromeOptions: {debuggerAddress: localhost:42205}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: linux, proxy: Proxy(), se:cdp: ws://localhost:42205/devtoo..., se:cdpVersion: 121.0.6167.139, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
Element: [[ChromeDriver: chrome on linux (828b2d6e917a0aca2f6e986d388255db)] -> css selector: div.active-choice:has([name='name'][value='PARAM3']) input[type='checkbox']]
Session ID: 828b2d6e917a0aca2f6e986d388255db
[INFO] 
[ERROR] Tests run: 77, Failures: 0, Errors: 1, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  07:59 min
[INFO] Finished at: 2024-02-06T21:39:31+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.2.5:test (default-test) on project uno-choice: 
[ERROR] 
[ERROR] Please refer to /home/kinow/Development/java/jenkins/active-choices-plugin/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.2.5:test (default-test) on project uno-choice: 

Please refer to /home/kinow/Development/java/jenkins/active-choices-plugin/target/surefire-reports for the individual test results.
Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:306)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:211)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:157)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:121)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:127)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.MojoFailureException: 

Please refer to /home/kinow/Development/java/jenkins/active-choices-plugin/target/surefire-reports for the individual test results.
Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
    at org.apache.maven.plugin.surefire.SurefireHelper.throwException (SurefireHelper.java:247)
    at org.apache.maven.plugin.surefire.SurefireHelper.reportExecution (SurefireHelper.java:159)
    at org.apache.maven.plugin.surefire.SurefirePlugin.handleSummary (SurefirePlugin.java:542)
    at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked (AbstractSurefireMojo.java:1102)
    at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute (AbstractSurefireMojo.java:910)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:301)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:211)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:165)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:157)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:121)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:127)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[ERROR] 
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

And env:

```bash
Apache Maven 3.8.5 (3599d3414f046de2324203b78ddcf9b5e4388aa0)
Maven home: /opt/apache-maven-3.8.5
Java version: 17.0.9, vendor: Private Build, runtime: /usr/lib/jvm/java-17-openjdk-amd64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.15.0-92-generic", arch: "amd64", family: "unix"
```

**Comment by kinow** [comment]
Well, solved the `master` test. I had to increase the dimension of the window to avoid elements being stacked on top of each other.

```diff
diff --git a/src/test/java/org/biouno/unochoice/UiAcceptanceTest.java b/src/test/java/org/biouno/unochoice/UiAcceptanceTest.java
index ff7c307..94cd6c2 100644
--- a/src/test/java/org/biouno/unochoice/UiAcceptanceTest.java
+++ b/src/test/java/org/biouno/unochoice/UiAcceptanceTest.java
@@ -70,7 +70,7 @@ public class UiAcceptanceTest {
             driver = new ChromeDriver(new ChromeOptions());
         }
         wait = new WebDriverWait(driver, MAX_WAIT);
-        driver.manage().window().setSize(new Dimension(1440, 900));
+        driver.manage().window().setSize(new Dimension(2560, 1440));
     }
 
     @After
```

Less one thing blocking this PR.

**Comment by kinow** [comment]
Rebasing `master` on this branch, and applying the diff above (which was working for me), and running `mvn clean test` on this branch, I get:

```bash
...
[INFO] Running org.biouno.unochoice.util.TestUtils
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 14.37 s -- in org.biouno.unochoice.util.TestUtils
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestForNodeLabelParameter.testNodeLabelParameterValueFound:143 Wrong number of HTML options rendered expected:<1> but was:<0>
[ERROR]   TestGroovyScriptParametersOrder.testGroovyScriptParametersOrder:145 Wrong number of HTML options rendered expected:<6> but was:<1>
[ERROR]   TestParametersXssVulnerabilities.testChoicesParameterXss:97 XSS string was not escaped!
[ERROR] Errors: 
[ERROR]   UiAcceptanceTest.test:119->checkOptions:252 » Timeout Expected condition failed: org.biouno.unochoice.UiAcceptanceTest$$Lambda$1191/0x0000000100a450b0@2a08c625 should have had [A1, A2, A3]. Had [] (tried for 60 second(s) with 500 milliseconds interval)
Build info: version: '4.17.0', revision: 'e52b1be057*'
System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.15.0-92-generic', java.version: '17.0.9'
Driver info: org.openqa.selenium.chrome.ChromeDriver
Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 121.0.6167.139, chrome: {chromedriverVersion: 121.0.6167.85 (3f98d690ad7e..., userDataDir: /tmp/.org.chromium.Chromium...}, fedcm:accounts: true, goog:chromeOptions: {debuggerAddress: localhost:43051}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: linux, proxy: Proxy(), se:cdp: ws://localhost:43051/devtoo..., se:cdpVersion: 121.0.6167.139, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
Session ID: aeae9b735e7507a53f1dea8d902ea501
[ERROR]   TestDynamicReferenceXss.testChoicesParameterXss:93 » Script TypeError: Cannot call method "getAttribute" of null (script in http://localhost:32793/jenkins/job/test0/build?delay=0sec from (60, 825) to (71, 12)#69)
[ERROR]   TestXssParameterNameBrowserConsole.testDynamicReferenceParameterXss:159 » Script TypeError: Cannot call method "getAttribute" of null (script in http://localhost:37643/jenkins/job/test0/build?delay=0sec from (60, 1104) to (71, 12)#69)
[INFO] 
[ERROR] Tests run: 77, Failures: 3, Errors: 3, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  08:32 min
[INFO] Finished at: 2024-02-06T22:30:12+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.2.5:test (default-test) on project uno-choice: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/kinow/Development/java/jenkins/active-choices-plugin/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
```

A bit late to continue troubleshooting it, but at least made some progress, and leaving these traces for others/myself-another-day to investigate it :+1: 

-Bruno

**Comment by kinow** [comment]
Recreated the scenario of one of the failing tests, and there are JS errors in the console. This `.split` function might be the one added here (that would make sense at least).

image

So I think the JS code added here, while simple, needs some rework to handle some corner cases when elements are not yet rendered, or have been updated.

I have to work on other active-choices issues, and triage tap-plugin issues too, so it might take me some time until I come back to this one. But if anyone manages to fix the code so that all tests pass, it would be a lot simpler to review & release this one :+1: - which would be great as this is a good feature for users.

Cheers,

Bruno

**Comment by c3p0-maif** [comment]
Hi @kinow ,
I've made smalls modifications in order to securise the spinner interactions and to, as suggested, disable the component during the loading.
Tests have been fixed ✌️

**Comment by c3p0-maif** [comment]
Spinner is now displayed in an absolute div so visibility toggle doesn't impact the page layout.

File: src/main/resources/org/biouno/unochoice/stapler/unochoice/UnoChoice.es6
Comment (kinow) [inline src/main/resources/org/biouno/unochoice/stapler/unochoice/UnoChoice.es6:525]: s/Active/Activate?

File: src/main/resources/org/biouno/unochoice/stapler/unochoice/UnoChoice.es6
Comment (kinow) [inline src/main/resources/org/biouno/unochoice/stapler/unochoice/UnoChoice.es6:447]: `let`?

**Review by kinow — COMMENTED** [review]
Thanks for updating the PR and fixing the tests, @c3p0-maif ! I had a quick look and the code looks good. Left a few comments, but nothing important. I will try to test it tomorrow, and if that works we should be ready to merge and release it! :muscle:

File: src/test/java/org/biouno/unochoice/jenkins_cert_1954/TestParametersXssVulnerabilities.java
Comment (kinow) [inline src/test/java/org/biouno/unochoice/jenkins_cert_1954/TestParametersXssVulnerabilities.java:107]: Thought it would be quite slow, but I think ~11 seconds on my old laptop should be fine for this acceptance test. We can improve it later if necessary. A good workaround :+1: 

```bash
[INFO] Running org.biouno.unochoice.jenkins_cert_1954.TestParametersXssVulnerabilities
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 11.86 s -- in org.biouno.unochoice.jenkins_cert_1954.TestParametersXssVulnerabilities
```

**Review by kinow — APPROVED** [review]
Tested and it seems to work as expected (cc @imoutsatsos something interesting to try with the next release).

output

@c3p0-maif I tested your branch after rebasing and squashing your commits to a single one. I will rebase squashing commits as I can see you clearly named each commit for code reviews, tests, etc (thanks). If the tests pass after doing that, it will be merged and included in the next release! (Which I will cut right away if there are no pending PR's or blockers, stay tuned!)

Thanks

**Pull request merged**
